### PR TITLE
Sa 3140 set new policy functions

### PR DESCRIPTION
**Pull request by jworkmanjc**
## Issues
* [SA-3143](https://jumpcloud.atlassian.net/browse/SA-3143) - Set/New-JCPolicy 

## What does this solve?

Two new functions, Set/New-JCPolicy created to help administrators programmatically set policies from the JumpCloud PowerShell module

## Is there anything particularly tricky?

Yes, some functionality must be tested by hand.

## How should this be tested?

Manual Test Cases are defined in the `Set-JCPolicy.tests.ps1` and `Set-JCPolicy.tests.ps1` files under the skipped 'Manual Test Cases' Context. These tests should be executed manually before approving the PR.

## Screenshots

[SA-3143]: https://jumpcloud.atlassian.net/browse/SA-3143?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: PowerShell/JumpCloud Module/JumpCloud.psm1 [hunk 1]
```diff
@@ -52,52 +52,22 @@ If ($PSVersionTable.PSEdition -eq 'Core') {
 # Set function aliases
 Set-Alias -Name:('New-JCAssociation') -Value:('Add-JCAssociation')
 # Argument completers
-# Populate values for function parameters. "Dynamic ValidateSet"
+# set Argument Completer(s) which no not require authentication
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/JumpCloud.psm1:55]: Just a note here, the filters in our documentation were not always correct, I've taken the argument completer out of the equation here. An added benefit is that the module loads much faster now. 

TODO: Team to check that `Get-JCSystemInsight -Table *tab*` still shows valid tables

File: PowerShell/JumpCloud Module/Private/Policies/Dispatch-JCPolicyConfigField.ps1 [hunk 2]
```diff
@@ -0,0 +1,71 @@
+function Dispatch-JCPolicyConfigField {
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Private/Policies/Dispatch-JCPolicyConfigField.ps1:1]: This function is not used, I think we can remove it

Comment (gweinjc) [reply PowerShell/JumpCloud Module/Private/Policies/Dispatch-JCPolicyConfigField.ps1:1]: Agreed

File: PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyConfigField.ps1 [hunk 3]
```diff
@@ -0,0 +1,57 @@
+function Get-JCPolicyConfigField {
```
[the remark this answers is not in the document]
Comment (gweinjc) [reply PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyConfigField.ps1:1]: Agreed

File: PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyTemplateConfigField.ps1 [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            # if ($field.DisplayOptions.Values.values)
+            if ($Field.DisplayType -eq 'select') {
+                $ValidationObject = @()
+                $validationRows = $field.DisplayOptions.select.Count
+                for ($i = 0; $i -lt $validationRows; $i++) {
+                    # Get Both Values:
+                    $val1 = $field.DisplayOptions.select[$i].text
+                    $val2 = $field.DisplayOptions.select[$i].value
+                    # Determine which one is an int
+                    # TODO: Joe to review here is this necessary for non-sdk calls?
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyTemplateConfigField.ps1:43]: Might need to clean up some of these comments/todos?

File: PowerShell/JumpCloud Module/Private/Policies/New-CustomRegistryTableRow.ps1 [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        $QWORD = New-Object System.Management.Automation.Host.ChoiceDescription "&QWORD", "QWORD"
+        $SZ = New-Object System.Management.Automation.Host.ChoiceDescription "&SZ", "String"
+        $Options = [System.Management.Automation.Host.ChoiceDescription[]]($DWORD, $EXPAND_SZ, $MULTI_SZ, $QWORD, $SZ)
+    }
+    process {
+        if (-Not $customData) {
+            $customData = (Read-Host "Please enter the registry data value")
+        }
+        if (-Not $customRegType) {
+            # $validRegTypes | Format-Table -Property Row, Name | Out-Host
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Private/Policies/New-CustomRegistryTableRow.ps1:59]: cleanup comment

Comment (gweinjc) [inline PowerShell/JumpCloud Module/Private/Policies/New-CustomRegistryTableRow.ps1:59]:
```suggestion
```

Cleanup

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        # print policy value
+        # if ('table' -in $templateObject.type) {
+        #     # If custom registry table, display all rows
+        #     $policyValues | Format-Table -Property @{label = "Row"; Expression = { $policyvalues.indexOf($_) } }, * | Out-Host
+        # } else {
+        #     # If not table, display just the row #, the name and value pairs
+        #     $templateobject | Format-Table @{label = "Row"; expression = { $TemplateObject.position -replace '\..+' } }, @{label = "Value"; expression = { $PolicyValues[$($TemplateObject.position) - 1].value } }, Label | Out-Host
+        #     # $templateObject | Format-Table @{label = "Row"; expression = { $TemplateObject.indexof($_) } }, @{label = "Value"; expression = { $PolicyValues[$TemplateObject.indexof($_)] } }, Label | Out-Host
+        # }
+
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1:48]:
```suggestion
```

Remove this block?

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        if ($fieldIndex) {
+            $field = $templateObject[$fieldIndex]
+            # $field = $templateObject | Where-Object { $_.position -eq $row }
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1:51]:
```suggestion
```

Remove?

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+                        try {
+                            if (($fieldValue -eq 't') -or ($fieldValue -eq 'true')) {
+                                $fieldValue = $true
+                            } elseif (($fieldValue -eq 'f') -or ($fieldValue -eq 'false')) {
+                                $fieldValue = $false
+                            } else {
+                                $fieldValue = [System.Convert]::ToBoolean($fieldValue)
+                            }
+                        } catch {
+                            Write-Warning "$fieldValue is not a boolean; please enter a boolean string value: (t/f) (true/false)"
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1:65]:
```suggestion
                            Write-Warning "$fieldValue is not a boolean; please enter a boolean value: (t/f) (true/false)"
```
Remove 'string' for clarity

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+
+                        }
+                    } While ($userChoice -ne 3)
+                    # finally update values
+                    $policyValues[$fieldIndex].value = $ValueObject
+                }
+                'file' {
+                    do {
+                        $filePath = Read-Host "Enter the file path location of the file for this policy"
+                        # write-host "testing pasth:$($filePath):"
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1:192]:
```suggestion
```

Cleanup comment

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+                                for ($i = 0; $i -lt $ValueObject.Count; $i++) {
+                                    $valueDisplay.Add([PSCustomObject]@{
+                                            row   = $i
+                                            value = $ValueObject[$i]
+                                        }) | Out-Null
+                                }
+                                $valueDisplay | Format-Table row, @{Label = "customData"; Expression = { $_.value | Select-Object -ExpandProperty customData } }, @{Label = "customRegType"; Expression = { $_.value | Select-Object -ExpandProperty customRegType } }, @{Label = "customLocation"; Expression = { $_.value | Select-Object -ExpandProperty customLocation } }, @{Label = "customValueName"; Expression = { $_.value | Select-Object -ExpandProperty customValueName } } | Out-Host
+                                $userChoice = $host.ui.PromptForChoice($title, $message, $options, 0)
+                            } else {
+                                # No values, just prompt to enter new value (todo, skip to entering new value)
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1:253]:
```suggestion
                                # No values, just prompt to enter new value
```

cleanup todo

File: PowerShell/JumpCloud Module/Public/Policies/New-JCPolicy.ps1 [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+                throw "template list missing; have you run Connect-JCOnline"
+            }
+            $templateObject = Get-JCPolicyTemplateConfigField -templateID $matchedTemplate
+            if ([String]::IsNullOrEmpty($templateObject.defaultName)) {
+                throw "Could not find policy template by specified Name"
+            }
+        }
+
+        if ($templateObject.objectMap -And ($TemplateName -OR $TemplateID)) {
+            # $templateObject = Get-JCPolicyTemplateConfigField -templateID $templateID
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Public/Policies/New-JCPolicy.ps1:52]:
```suggestion
```

Cleanup

File: PowerShell/JumpCloud Module/Public/Policies/New-JCPolicy.ps1 [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+                                $templateObject.objectMap[$i].value = $($keyValue)
+                            }
+                        }
+                        'table' {
+                            # For custom registry table, validate the object
+                            if ($templateObject.objectMap[$i].configFieldName -eq "customRegTable") {
+                                # TODO: validate that the objects passed in are valid
+                                # if ($keyValue.getType().Name -ne 'Object[]') {
+                                #     throw "The object passed in as values input does not match the expected value type. Data is expected to be formatted as a hashtable: @{customData = 'someString'; customLocation = 'location'; customRegType = 'DWORD'; customValueName = 'registryKeyValue' } or ArrayList of PSCustomObjects"
+                                # }
```
Comment (gweinjc) [inline PowerShell/JumpCloud Module/Public/Policies/New-JCPolicy.ps1:177]: Does this TODO still need to be done?

**Review by kmaranionjc — CHANGES_REQUESTED** [review]

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 (hunk 8 above)
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1]: good catch, removed

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 (hunk 7 above)
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1]: removed

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 (hunk 6 above)
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1]: removed

File: PowerShell/JumpCloud Module/Private/Policies/New-CustomRegistryTableRow.ps1 (hunk 5 above)
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/New-CustomRegistryTableRow.ps1]: removed

File: PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyTemplateConfigField.ps1 (hunk 4 above)
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyTemplateConfigField.ps1]: removed

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 (hunk 10 above)
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1]: removed

File: PowerShell/JumpCloud Module/Private/Policies/Dispatch-JCPolicyConfigField.ps1 [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                } catch {
+                    #TODO: throw generic message
+                }
+            }
+
+        } until (($rowPrompt -eq 'c'))
+    }
+    end {
+        return $policyValues
+        # TODO: return the new policy values here:
```
[the remark this answers is not in the document]
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Dispatch-JCPolicyConfigField.ps1]: removd!

File: PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyConfigField.ps1 [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        }
+        # if ($policyValues) {
+        # }
+
+
+    }
+    end {
+        # Return template config field
+        return $returnObject
+        # TODO: return the whole enchilada
```
[the remark this answers is not in the document]
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyConfigField.ps1]: removed

File: PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyTemplateConfigField.ps1 [hunk 15]
```diff
@@ -0,0 +1,84 @@
+function Get-JCPolicyTemplateConfigField {
+    # TemplateID of the policy to be modified
+    [CmdletBinding()]
+    param (
+        [Parameter(mandatory = $true)]
+        [System.String]
+        $templateID
+    )
+    begin {
+        # TODO: validate templateID
```
[the remark this answers is not in the document]
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Get-JCPolicyTemplateConfigField.ps1]: removed

File: PowerShell/JumpCloud Module/Public/Policies/New-JCPolicy.ps1 (hunk 11 above)
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Public/Policies/New-JCPolicy.ps1]: removed

File: PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1 (hunk 9 above)
Comment (jworkmanjc) [reply PowerShell/JumpCloud Module/Private/Policies/Set-JCPolicyConfigField.ps1]: removed

**Review by kmaranionjc — APPROVED** [review]

**Review by gweinjc — APPROVED** [review]

**Pull request merged**